tools/clang/lib/SPIRV/SpirvModule.cpp

```diff
@@ -459,63 +249,188 @@ void SpirvModule::addSource(SpirvSource *src) {
+    assert(inst != nullptr);
+    if (!visitor(inst->getElementType()))
+      return;
+  }
+    return;
+  case SpirvInstruction::IK_DebugTypeVector: {
+    SpirvDebugTypeVector *inst = dyn_cast<SpirvDebugTypeVector>(di);
+    assert(inst != nullptr);
+    if (!visitor(inst->getElementType()))
+      return;
+  }
+    return;
+  case SpirvInstruction::IK_DebugTypeFunction: {
+    SpirvDebugTypeFunction *inst = dyn_cast<SpirvDebugTypeFunction>(di);
+    assert(inst != nullptr);
+    if (!visitor(inst->getReturnType()))
+      return;
+    for (auto *param : inst->getParamTypes())
+      if (!visitor(param))
+        return;
+  }
+    return;
+  case SpirvInstruction::IK_DebugTypeComposite: {
+    SpirvDebugTypeComposite *inst = dyn_cast<SpirvDebugTypeComposite>(di);
+    assert(inst != nullptr);
+    if (!visitor(inst->getSource()))
+      return;
+    if (!visitor(inst->getDebugInfoNone()))
+      return;
+  }
+    return;
+  case SpirvInstruction::IK_DebugTypeMember: {
+    SpirvDebugTypeMember *inst = dyn_cast<SpirvDebugTypeMember>(di);
+    assert(inst != nullptr);
+    if (!visitor(inst->getType()))
+      return;
+    if (!visitor(inst->getSource()))
+      return;
+  }
+    return;
+  case SpirvInstruction::IK_DebugTypeTemplate: {
+    SpirvDebugTypeTemplate *inst = dyn_cast<SpirvDebugTypeTemplate>(di);
+    assert(inst != nullptr);
+    for (auto *param : inst->getParams())
+      if (!visitor(param))
+        return;
+    if (!visitor(inst->getTarget()))
+      return;
+  }
+    return;
+  case SpirvInstruction::IK_DebugTypeTemplateParameter: {
+    SpirvDebugTypeTemplateParameter *inst =
+        dyn_cast<SpirvDebugTypeTemplateParameter>(di);
+    assert(inst != nullptr);
+    if (!visitor(inst->getActualType()))
+      return;
+    if (!visitor(inst->getSource()))
+      return;
+  }
+    return;
+  default:
+    return;
+  }
 }
 
-void SpirvModule::addDebugInfo(SpirvDebugInstruction *info) {
-  assert(info);
-  debugInfo.push_back(info);
-}
+void SpirvModule::sortDebugInstructionsInPostOrder() {
+  // Collect nodes without predecessor.
+  llvm::SmallSet<SpirvDebugInstruction *, 32> visited;
+  for (auto *di : debugInstructions) {
+    whileEachOperandOfDebugInstruction(
+        di, [&visited](SpirvDebugInstruction *operand) {
+          if (operand != nullptr)
+            visited.insert(operand);
+          return true;
+        });
+  }
+  llvm::SmallVector<SpirvDebugInstruction *, 32> stack;
+  for (auto *di : debugInstructions) {
+    if (visited.count(di) == 0)
+      stack.push_back(di);
+  }
 
-void SpirvModule::addModuleProcessed(SpirvModuleProcessed *p) {
-  assert(p);
-  moduleProcesses.push_back(p);
+  // Sort debug instructions in a post order. We puts successors in the first
+  // places of `debugInstructions`. For example, `DebugInfoNone` does not have
+  // any operand, which means it does not have any successors. We have to place
+  // it earlier than the instructions using it.
+  debugInstructions.clear();
```

Comment:
You are clearing a vector of pointers and then re-populating the vector. I highly suggest getting the size of `debugInstructions` before clearing it, and then after the while loop add an `assert` to make sure the new size is equal to the old size. Otherwise you have lost pointers to memory forever (leak).

In addition to leaks, not having the assert could result in really-hard-to-find bugs. For example if `whileEachOperandOfDebugInstruction` is missing a debug instruction `case` statement or if it is missing an operand of some debug instruction, you could have a leak, which can be caught by the assert.

Reply:
Oh nice suggestion